### ISSUE TITLE
Async with Task<T>

### ISSUE THREAD

**Issue by jayschroeder** [mattflo/NSpec#122]
First, still learning the TDD / BDD mindset and I'm really liking NSpec, thanks for your time and effort.

The documentation isn't real clear as to how the runner handles hooks marked with Async (beforeAsync, actAsync, ...). Does the runner wait for task completion so that "it" specs properly (ex: "beforeAsync" makes sure its operation is complete so "it" specs can evaluate results)?

I'm using `Task<T>` but it causes a runtime error (below). From the error message it appears the issue is the runner can only handle Task (not the generic version)?

```
Exception executing context: nspec. Verify list. given 1 valid switch
System.Reflection.TargetInvocationException: Exception has been thrown by the ta
rget of an invocation. ---> System.MissingMethodException: Method not found: 'Vo
id NSpec.nspec.set_beforeAsync(System.Func`1<System.Threading.Tasks.Task>)'.
```

Jay

**Comment by BrainCrumbz** [comment]
Thanks for your appreciation! Yes ... docs on async support are still a little sparse.

I take it you're referring to class/method level hooks: yes, you got that right. For them it only makes sense to return `Task`, not `Task<T>`. If you think about it, they're the async *"evolution"* of sync hooks, which always return `void`, no actual value returned. So they only need to cope with the async nature of code, again, no actual value to return from Task.

In order for async parts of code to work correctly, you should put them inside async hooks (e.g. `async Task before_each()`, `beforeAsync`) or async specs (e.g `async Task it_should_xyz()`, `itAsync`).

**Comment by jayschroeder** [comment]
So I'm a little confused. Let's try this: if I were only returning a `Task` (to avoid the run-time error) does  the runner wait on any of the "sync hook" method that contain an `await` until they complete before it executes the next level of hook methods in the hierarchy? Ex: the runner will ensure the `beforeAsync` is complete before executing any `it` hooks?

Based on your response I think the answer is no. If that's the case then what do the "sync hooks" actually do?

**Comment by BrainCrumbz** [comment]
Sorry about confusion. I don't get exactly what you mean with a *sync hook method that contain an await*. In order to include an `await` in its body and have everything working correctly, the hook method must be flagged as `async` and return a `Task` in its signature, as in the following:

```cs
public class an_example_with_async_hooks_at_class_level : nspec
{
  async Task before_each()
  {
    // ...
    await SetupScenarioAsync();
    // ...
  }

  void it_should_do_something()
  {
    DoSomething();
  }

  async Task it_should_do_something_async()
  {
    await DoSomethingAsync();
  }
}
```

Doing so ensures that async work within `before_each()` will be awaited for, and only then further hooks or specs will be processed.  
The same scenario, when working at context level, goes like:

```cs
public class an_example_with_async_hooks_at_context_level : nspec
{
  // ...
  void given_some_context()
  {
    // note the 'Async' suffix

    beforeAsync = async () => await SetupScenarioAsync();

    it["should do something"] = () => DoSomething();

    itAsync["should do something async"] = async () => await DoSomethingAsync();
  }
  // ...
}
```

As you can see there, you still have to flag lamdba expression as `async` in order to let it contain an `await`. Only the `Task` is not explicit, just because that's an expression, not a method.

The "sync" hooks and "sync" specs are totally fine to use when you have your unit under test does not expose any async code to check. So if when setting up your test scenario and when checking some method under test no async code is involved, you can (should) use sync hooks and specs.

Maybe you can provide a code snippet to help me understand your scenario.

**Comment by amirrajan** [comment]
Poking myself: https://github.com/mattflo/NSpec/issues/91

**Comment by jayschroeder** [comment]
Sorry for the delayed response. Thanks, your reply has cleared up my confusion. I did try a small sample program and could clearly see the difference between using asynchronous hook (ex: beforeAsync vs before) with the latter causing a failed spec due to a race (spec statement executes before await is completed) while the first resulted in successful spec. I also tried a task with a return value and it worked so I must have had some other problem in my first post. I'll have to go back and check my real project to see if I can find my problem.

**Comment by BrainCrumbz** [comment]
Don't take my word for it, just going on top of my head: you get an error for class-level hooks (`async Task before_each() { ... }`), but probably not for context-level hooks (`beforeAsync = async () => await ...`), where the presence of return value/type is ignored.

**Comment by jayschroeder** [comment]
Ok, I finally got to the bottom of my problem (and I'm embarrassed). I now recall when I first looked at NSpec I did not understanding nuget, so I did what was familiar - download the zip, extract to a directory and I also added an External Tool menu item to VS to call the runner. I read up on nuget and tried it on my home computer until I felt comfortable to do the same on my work computer. Unfortunately, my project was compiling against the latest version dll but I forgot to change the External tool menu so it was calling an earlier version of the runner. What's interesting is that I didn't see any problems until I introduced async code. All is well now, works great any my test code is far more understandable. Thanks again for your time.

**Comment by BrainCrumbz** [comment]
Good to know! When you speak about *the runner* being called from external tool item I guess you mean `NSpecRunner.exe`, as mentioned on [nspec Home](http://nspec.org/).

Just as blatant self-advertising, we released a VS Test Runner for NSpec in the open: [NSpec.VsAdapter](https://github.com/BrainCrumbz/NSpec.VsAdapter). Actually that has been recently improved to support async tests as well. HTH

**Comment by jayschroeder** [comment]
Yes. External Tools is simply a user defined menu under the Tools menu. I set it up to call the NSpecRunner.exe and let VS build the name of the test project to pass as a parameter. I'm using the express edition so I can use the VS Test Runner stuff (I tried to setup NUnit without success, but I now prefer nspec).

Jay

**Comment by BrainCrumbz** [comment]
Not sure you're bound to some dev environment constraint, but maybe you can give a go to VS Community Edition (2013 or 2015), as those are more powerful than Express editions.
